File: Leetcode/src/com/leetcode/matrix/_59_Spiral_Matrix_II.py
'''
Date: 12/07/2020 09:42:28
LastEditTime: 12/07/2020 09:43:42
Description: Simulation, Matrix
'''


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        res = [[0 for _ in range(n)] for _ in range(n)]
        row_begin, row_end = 0, n - 1
        col_begin, col_end = 0, n - 1
        number = 1
        while row_begin <= row_end and col_begin <= col_end:
            for j in range(col_begin, col_end + 1):
                res[row_begin][j] = number
                number += 1
            row_begin += 1
            if row_begin > row_end:
                break

            for i in range(row_begin, row_end + 1):
                res[i][col_end] = number
                number += 1
            col_end -= 1
            if col_end < col_begin:
                break

            for j in range(col_end, col_begin - 1, -1):
                res[row_end][j] = number
                number += 1
            row_end -= 1
            if row_end < row_begin:
                break

            for i in range(row_end, row_begin - 1, -1):
                res[i][col_begin] = number
                number += 1
            col_begin += 1
            if col_begin > col_end:
                break
        return res

File: Leetcode/src/com/leetcode/matrix/59_spiral_matrix_ii/_59_SpiralMatrixII.java
package com.leetcode.matrix;

public class _59_SpiralMatrixII {

    /**
     * 59. Spiral Matrix II
     * When: 2019/05/29
     * Review1:2019/8/5
     * solution:
     * 跟之前那道题一样的思路，只是做了一个值来循环
     * @param n
     * @return
     */
    //time: O(m * n) space: O(m * n)
    public int[][] generateMatrix(int n) {
        int[][] res = new int[n][n];
        int start = 1;
        int rowBegin = 0;
        int rowEnd = n - 1;
        int colBegin = 0;
        int colEnd = n - 1;
        while (rowBegin <= rowEnd && colBegin <= colEnd) {
            //向右
            for (int i = colBegin; i <= colEnd; i++) {
                res[rowBegin][i] = start;
                start++;
            }
            rowBegin++;
            //向下
            for (int i = rowBegin; i <= rowEnd; i++) {
                res[i][colEnd] = start;
                start++;
            }
            colEnd--;

            //向左
            if (rowBegin <= rowEnd) {
                for (int i = colEnd; i >= colBegin; i--) {
                    res[rowEnd][i] = start;
                    start++;
                }
            }
            rowEnd--;

            //向上
            if (colBegin <= colEnd) {
                for (int i = rowEnd; i >= rowBegin; i--) {
                    res[i][colBegin] = start;
                    start++;
                }
            }
            colBegin++;
        }
        return res;
    }
}
